File: src/services/memory.ts
/**
 * Async wrapper for the HEBBS MemoryService gRPC methods.
 */

import type { Metadata, ServiceError } from '@grpc/grpc-js';
import { mapGrpcError } from '../errors.js';
import {
  grpcUnary,
  toProtoStruct,
  protoToMemory,
  protoToRecallResult,
  edgeToProto,
  scoringWeightsToProto,
  strategyConfigToProto,
  strategyToProto,
  strategyFromProto,
} from '../proto.js';
import type {
  Memory,
  RecallOutput,
  PrimeOutput,
  ForgetResult,
  ScoringWeights,
  RecallStrategyConfig,
  Edge,
  StrategyError,
} from '../types.js';

/* eslint-disable @typescript-eslint/no-explicit-any */

export class MemoryService {
  constructor(
    private readonly stub: any,
    private readonly metadata: Metadata,
    private readonly tenantId?: string,
  ) {}

  async remember(
    content: string,
    importance?: number,
    context?: Record<string, unknown>,
    entityId?: string,
    edges?: Edge[],
  ): Promise<Memory> {
    const req: any = { content };
    if (importance !== undefined) req.importance = importance;
    const ctx = toProtoStruct(context);
    if (ctx) req.context = ctx;
    if (entityId) req.entityId = entityId;
    if (this.tenantId) req.tenantId = this.tenantId;
    if (edges && edges.length > 0) {
      req.edges = edges.map(edgeToProto);
    }

    try {
      const resp = await grpcUnary<any>((cb) =>
        this.stub.remember(req, this.metadata, cb),
      );
      return protoToMemory(resp.memory);
    } catch (e) {
      throw mapGrpcError(e);
    }
  }

  async get(memoryId: Buffer): Promise<Memory> {
    const req: any = { memoryId };
    if (this.tenantId) req.tenantId = this.tenantId;

    try {
      const resp = await grpcUnary<any>((cb) =>
        this.stub.get(req, this.metadata, cb),
      );
      return protoToMemory(resp.memory);
    } catch (e) {
      throw mapGrpcError(e);
    }
  }

  async recall(
    cue: string,
    strategies?: (string | RecallStrategyConfig)[],
    topK?: number,
    entityId?: string,
    scoringWeights?: ScoringWeights,
    cueContext?: Record<string, unknown>,
  ): Promise<RecallOutput> {
    const stratConfigs: any[] = [];
    for (const s of strategies ?? ['similarity']) {
      if (typeof s === 'string') {
        const cfg: any = { strategyType: strategyToProto(s) };
        if (entityId) cfg.entityId = entityId;
        stratConfigs.push(cfg);
      } else {
        stratConfigs.push(strategyConfigToProto(s, entityId));
      }
    }

    const req: any = { cue, strategies: stratConfigs };
    if (topK !== undefined) req.topK = topK;
    if (this.tenantId) req.tenantId = this.tenantId;
    if (scoringWeights) {
      req.scoringWeights = scoringWeightsToProto(scoringWeights);
    }
    const ctx = toProtoStruct(cueContext);
    if (ctx) req.cueContext = ctx;

    try {
      const resp = await grpcUnary<any>((cb) =>
        this.stub.recall(req, this.metadata, cb),
      );
      const results = (resp.results ?? []).map(protoToRecallResult);
      const strategyErrors: StrategyError[] = (
        resp.strategyErrors ?? []
      ).map((se: any) => ({
        strategy: strategyFromProto(se.strategy ?? se.strategyType ?? 0),
        message: se.message ?? '',
      }));
      return { results, strategyErrors };
    } catch (e) {
      throw mapGrpcError(e);
    }
  }

  async prime(
    entityId: string,
    maxMemories?: number,
    similarityCue?: string,
    scoringWeights?: ScoringWeights,
  ): Promise<PrimeOutput> {
    const req: any = { entityId };
    if (maxMemories !== undefined) req.maxMemories = maxMemories;
    if (similarityCue) req.similarityCue = similarityCue;
    if (this.tenantId) req.tenantId = this.tenantId;
    if (scoringWeights) {
      req.scoringWeights = scoringWeightsToProto(scoringWeights);
    }

    try {
      const resp = await grpcUnary<any>((cb) =>
        this.stub.prime(req, this.metadata, cb),
      );
      return {
        results: (resp.results ?? []).map(protoToRecallResult),
        temporalCount: resp.temporalCount ?? resp.temporal_count ?? 0,
        similarityCount: resp.similarityCount ?? resp.similarity_count ?? 0,
      };
    } catch (e) {
      throw mapGrpcError(e);
    }
  }

  async revise(
    memoryId: Buffer,
    content?: string,
    importance?: number,
    context?: Record<string, unknown>,
    entityId?: string,
  ): Promise<Memory> {
    const req: any = { memoryId };
    if (content !== undefined) req.content = content;
    if (importance !== undefined) req.importance = importance;
    const ctx = toProtoStruct(context);
    if (ctx) req.context = ctx;
    if (entityId) req.entityId = entityId;
    if (this.tenantId) req.tenantId = this.tenantId;

    try {
      const resp = await grpcUnary<any>((cb) =>
        this.stub.revise(req, this.metadata, cb),
      );
      return protoToMemory(resp.memory);
    } catch (e) {
      throw mapGrpcError(e);
    }
  }

  async forget(
    entityId?: string,
    memoryIds?: Buffer[],
  ): Promise<ForgetResult> {
    const req: any = {};
    if (entityId) req.entityId = entityId;
    if (memoryIds && memoryIds.length > 0) req.memoryIds = memoryIds;
    if (this.tenantId) req.tenantId = this.tenantId;

    try {
      const resp = await grpcUnary<any>((cb) =>
        this.stub.forget(req, this.metadata, cb),
      );
      return {
        forgottenCount: resp.forgottenCount ?? resp.forgotten_count ?? 0,
        cascadeCount: resp.cascadeCount ?? resp.cascade_count ?? 0,
        tombstoneCount: resp.tombstoneCount ?? resp.tombstone_count ?? 0,
        truncated: resp.truncated ?? false,
      };
    } catch (e) {
      throw mapGrpcError(e);
    }
  }

  async setPolicy(
    maxSnapshotsPerMemory?: number,
    autoForgetThreshold?: number,
    decayHalfLifeDays?: number,
  ): Promise<boolean> {
    const req: any = {};
    if (this.tenantId) req.tenantId = this.tenantId;
    if (maxSnapshotsPerMemory !== undefined)
      req.maxSnapshotsPerMemory = maxSnapshotsPerMemory;
    if (autoForgetThreshold !== undefined)
      req.autoForgetThreshold = autoForgetThreshold;
    if (decayHalfLifeDays !== undefined)
      req.decayHalfLifeDays = decayHalfLifeDays;

    try {
      const resp = await grpcUnary<any>((cb) =>
        this.stub.setPolicy(req, this.metadata, cb),
      );
      return resp.applied ?? false;
    } catch (e) {
      throw mapGrpcError(e);
    }
  }
}

/* eslint-enable @typescript-eslint/no-explicit-any */

File: src/hebbs/services/memory.py
"""Async wrapper for the HEBBS MemoryService gRPC methods."""

from __future__ import annotations

from typing import Any

from google.protobuf import struct_pb2

from hebbs._generated import hebbs_pb2, hebbs_pb2_grpc
from hebbs.exceptions import _map_grpc_error
from hebbs.types import (
    Edge,
    EdgeType,
    ForgetResult,
    Memory,
    MemoryKind,
    PrimeOutput,
    RecallOutput,
    RecallResult,
    RecallStrategy,
    RecallStrategyConfig,
    ScoringWeights,
    StrategyDetail,
    StrategyError,
)

_EDGE_TYPE_MAP = {
    EdgeType.CAUSED_BY: hebbs_pb2.EDGE_TYPE_CAUSED_BY,
    EdgeType.RELATED_TO: hebbs_pb2.EDGE_TYPE_RELATED_TO,
    EdgeType.FOLLOWED_BY: hebbs_pb2.EDGE_TYPE_FOLLOWED_BY,
    EdgeType.REVISED_FROM: hebbs_pb2.EDGE_TYPE_REVISED_FROM,
    EdgeType.INSIGHT_FROM: hebbs_pb2.EDGE_TYPE_INSIGHT_FROM,
}

_STRATEGY_MAP = {
    "similarity": hebbs_pb2.SIMILARITY,
    "temporal": hebbs_pb2.TEMPORAL,
    "causal": hebbs_pb2.CAUSAL,
    "analogical": hebbs_pb2.ANALOGICAL,
}

_MEMORY_KIND_REVERSE: dict[int, MemoryKind] = {
    hebbs_pb2.MEMORY_KIND_EPISODE: MemoryKind.EPISODE,
    hebbs_pb2.MEMORY_KIND_INSIGHT: MemoryKind.INSIGHT,
    hebbs_pb2.MEMORY_KIND_REVISION: MemoryKind.REVISION,
}

_STRATEGY_REVERSE: dict[int, str] = {
    hebbs_pb2.SIMILARITY: "similarity",
    hebbs_pb2.TEMPORAL: "temporal",
    hebbs_pb2.CAUSAL: "causal",
    hebbs_pb2.ANALOGICAL: "analogical",
}


def _to_proto_scoring_weights(sw: ScoringWeights | dict) -> hebbs_pb2.ScoringWeights:
    if isinstance(sw, dict):
        sw = ScoringWeights(**{k: v for k, v in sw.items() if v is not None})
    proto = hebbs_pb2.ScoringWeights(
        w_relevance=sw.w_relevance,
        w_recency=sw.w_recency,
        w_importance=sw.w_importance,
        w_reinforcement=sw.w_reinforcement,
    )
    if sw.max_age_us is not None:
        proto.max_age_us = sw.max_age_us
    if sw.reinforcement_cap is not None:
        proto.reinforcement_cap = sw.reinforcement_cap
    return proto


def _dict_to_struct(d: dict[str, Any] | None) -> struct_pb2.Struct | None:
    if not d:
        return None
    s = struct_pb2.Struct()
    s.update(d)
    return s


def _struct_to_dict(s: Any) -> dict[str, Any]:
    if s is None or not s.fields:
        return {}
    from google.protobuf.json_format import MessageToDict
    return MessageToDict(s)


def _proto_to_memory(m: Any) -> Memory:
    return Memory(
        id=bytes(m.memory_id),
        content=m.content,
        importance=m.importance,
        context=_struct_to_dict(m.context),
        entity_id=m.entity_id if m.entity_id else None,
        created_at=m.created_at,
        updated_at=m.updated_at,
        last_accessed_at=m.last_accessed_at,
        access_count=m.access_count,
        decay_score=m.decay_score,
        kind=_MEMORY_KIND_REVERSE.get(m.kind, MemoryKind.UNSPECIFIED),
        embedding=list(m.embedding),
        source_memory_ids=[bytes(sid) for sid in m.source_memory_ids],
    )


def _proto_to_strategy_detail(sd: Any) -> StrategyDetail:
    return StrategyDetail(
        strategy=_STRATEGY_REVERSE.get(sd.strategy_type, "unknown"),
        relevance=sd.relevance,
        distance=sd.distance if sd.HasField("distance") else None,
        timestamp=sd.timestamp if sd.HasField("timestamp") else None,
        rank=sd.rank if sd.HasField("rank") else None,
        depth=sd.depth if sd.HasField("depth") else None,
        embedding_similarity=sd.embedding_similarity if sd.HasField("embedding_similarity") else None,
        structural_similarity=sd.structural_similarity if sd.HasField("structural_similarity") else None,
    )


def _proto_to_recall_result(r: Any) -> RecallResult:
    return RecallResult(
        memory=_proto_to_memory(r.memory),
        score=r.score,
        strategy_details=[_proto_to_strategy_detail(sd) for sd in r.strategy_details],
    )


def _build_strategy_config_proto(
    sc: RecallStrategyConfig,
    fallback_entity_id: str | None = None,
) -> hebbs_pb2.RecallStrategyConfig:
    st = _STRATEGY_MAP.get(sc.strategy, hebbs_pb2.UNSPECIFIED_STRATEGY)
    cfg = hebbs_pb2.RecallStrategyConfig(strategy_type=st)

    eid = sc.entity_id or fallback_entity_id
    if eid:
        cfg.entity_id = eid
    if sc.top_k is not None:
        cfg.top_k = sc.top_k
    if sc.ef_search is not None:
        cfg.ef_search = sc.ef_search
    if sc.time_range is not None:
        cfg.time_range.CopyFrom(
            hebbs_pb2.TimeRange(start_us=sc.time_range[0], end_us=sc.time_range[1])
        )
    if sc.seed_memory_id is not None:
        cfg.seed_memory_id = sc.seed_memory_id
    if sc.edge_types:
        for et in sc.edge_types:
            proto_et = _EDGE_TYPE_MAP.get(et, hebbs_pb2.EDGE_TYPE_UNSPECIFIED)
            cfg.edge_types.append(proto_et)
    if sc.max_depth is not None:
        cfg.max_depth = sc.max_depth
    if sc.analogical_alpha is not None:
        cfg.analogical_alpha = sc.analogical_alpha
    return cfg


class MemoryServiceClient:
    """Async client for the HEBBS MemoryService."""

    def __init__(self, stub: hebbs_pb2_grpc.MemoryServiceStub, tenant_id: str | None = None) -> None:
        self._stub = stub
        self._tenant_id = tenant_id

    async def remember(
        self,
        content: str,
        importance: float | None = None,
        context: dict[str, Any] | None = None,
        entity_id: str | None = None,
        edges: list[Edge] | None = None,
    ) -> Memory:
        req = hebbs_pb2.RememberRequest(content=content)
        if importance is not None:
            req.importance = importance
        ctx = _dict_to_struct(context)
        if ctx is not None:
            req.context.CopyFrom(ctx)
        if entity_id:
            req.entity_id = entity_id
        if self._tenant_id:
            req.tenant_id = self._tenant_id
        if edges:
            for e in edges:
                proto_edge = hebbs_pb2.Edge(
                    target_id=e.target_id,
                    edge_type=_EDGE_TYPE_MAP.get(e.edge_type, hebbs_pb2.EDGE_TYPE_UNSPECIFIED),
                )
                if e.confidence is not None:
                    proto_edge.confidence = e.confidence
                req.edges.append(proto_edge)

        try:
            resp = await self._stub.Remember(req)
        except Exception as e:
            raise _map_grpc_error(e) from e
        return _proto_to_memory(resp.memory)

    async def get(self, memory_id: bytes) -> Memory:
        req = hebbs_pb2.GetRequest(memory_id=memory_id)
        if self._tenant_id:
            req.tenant_id = self._tenant_id
        try:
            resp = await self._stub.Get(req)
        except Exception as e:
            raise _map_grpc_error(e) from e
        return _proto_to_memory(resp.memory)

    async def recall(
        self,
        cue: str,
        strategies: list[str | RecallStrategyConfig] | None = None,
        top_k: int | None = None,
        entity_id: str | None = None,
        scoring_weights: ScoringWeights | dict | None = None,
        cue_context: dict[str, Any] | None = None,
    ) -> RecallOutput:
        strat_configs = []
        for s in (strategies or ["similarity"]):
            if isinstance(s, str):
                st = _STRATEGY_MAP.get(s, hebbs_pb2.UNSPECIFIED_STRATEGY)
                cfg = hebbs_pb2.RecallStrategyConfig(strategy_type=st)
                if entity_id:
                    cfg.entity_id = entity_id
            elif isinstance(s, RecallStrategyConfig):
                cfg = _build_strategy_config_proto(s, entity_id)
            else:
                raise TypeError(f"strategies must contain str or RecallStrategyConfig, got {type(s)}")
            strat_configs.append(cfg)

        req = hebbs_pb2.RecallRequest(cue=cue, strategies=strat_configs)
        if top_k is not None:
            req.top_k = top_k
        if self._tenant_id:
            req.tenant_id = self._tenant_id
        if scoring_weights is not None:
            req.scoring_weights.CopyFrom(_to_proto_scoring_weights(scoring_weights))
        if cue_context is not None:
            ctx = _dict_to_struct(cue_context)
            if ctx is not None:
                req.cue_context.CopyFrom(ctx)

        try:
            resp = await self._stub.Recall(req)
        except Exception as e:
            raise _map_grpc_error(e) from e

        return RecallOutput(
            results=[_proto_to_recall_result(r) for r in resp.results],
            strategy_errors=[
                StrategyError(
                    strategy=_STRATEGY_REVERSE.get(se.strategy, "unknown"),
                    message=se.message,
                )
                for se in resp.strategy_errors
            ],
        )

    async def prime(
        self,
        entity_id: str,
        max_memories: int | None = None,
        similarity_cue: str | None = None,
        scoring_weights: ScoringWeights | dict | None = None,
    ) -> PrimeOutput:
        req = hebbs_pb2.PrimeRequest(entity_id=entity_id)
        if max_memories is not None:
            req.max_memories = max_memories
        if similarity_cue:
            req.similarity_cue = similarity_cue
        if self._tenant_id:
            req.tenant_id = self._tenant_id
        if scoring_weights is not None:
            req.scoring_weights.CopyFrom(_to_proto_scoring_weights(scoring_weights))

        try:
            resp = await self._stub.Prime(req)
        except Exception as e:
            raise _map_grpc_error(e) from e

        return PrimeOutput(
            results=[_proto_to_recall_result(r) for r in resp.results],
            temporal_count=resp.temporal_count,
            similarity_count=resp.similarity_count,
        )

    async def revise(
        self,
        memory_id: bytes,
        content: str | None = None,
        importance: float | None = None,
        context: dict[str, Any] | None = None,
        entity_id: str | None = None,
    ) -> Memory:
        req = hebbs_pb2.ReviseRequest(memory_id=memory_id)
        if content is not None:
            req.content = content
        if importance is not None:
            req.importance = importance
        ctx = _dict_to_struct(context)
        if ctx is not None:
            req.context.CopyFrom(ctx)
        if entity_id:
            req.entity_id = entity_id
        if self._tenant_id:
            req.tenant_id = self._tenant_id

        try:
            resp = await self._stub.Revise(req)
        except Exception as e:
            raise _map_grpc_error(e) from e
        return _proto_to_memory(resp.memory)

    async def forget(
        self,
        entity_id: str | None = None,
        memory_ids: list[bytes] | None = None,
    ) -> ForgetResult:
        req = hebbs_pb2.ForgetRequest()
        if entity_id:
            req.entity_id = entity_id
        if memory_ids:
            req.memory_ids.extend(memory_ids)
        if self._tenant_id:
            req.tenant_id = self._tenant_id

        try:
            resp = await self._stub.Forget(req)
        except Exception as e:
            raise _map_grpc_error(e) from e

        return ForgetResult(
            forgotten_count=resp.forgotten_count,
            cascade_count=resp.cascade_count,
            tombstone_count=resp.tombstone_count,
            truncated=resp.truncated,
        )

    async def set_policy(
        self,
        max_snapshots_per_memory: int | None = None,
        auto_forget_threshold: float | None = None,
        decay_half_life_days: float | None = None,
    ) -> bool:
        req = hebbs_pb2.SetPolicyRequest()
        if self._tenant_id:
            req.tenant_id = self._tenant_id
        if max_snapshots_per_memory is not None:
            req.max_snapshots_per_memory = max_snapshots_per_memory
        if auto_forget_threshold is not None:
            req.auto_forget_threshold = auto_forget_threshold
        if decay_half_life_days is not None:
            req.decay_half_life_days = decay_half_life_days

        try:
            resp = await self._stub.SetPolicy(req)
        except Exception as e:
            raise _map_grpc_error(e) from e
        return resp.applied
